### ISSUE TITLE
ibm_db_execute: arrays of PYTHON_DECIMAL fails for some valid column-datatypes

### ISSUE THREAD

**Issue by imavo** [ibmdb/python-ibmdb#838]
- Operating System Name: linux x46 (or any)
- db2level output from clidriver if in use: 11.5.8.0 (or any)
- Target Db2 Server Version: 11.5.8.0
- Python Version: 3.9
- ibm_db version: 3.1.4
- For non-Windows, output of below commands:
  uname (n/a)
  uname -m (n/a)
- Value of below environment variables if set:
  IBM_DB_HOME: (not relevant)
  PATH:
  LIB/LD_LIBRARY_PATH/DYLD_LIBRARY_PATH: (not relevant)

- Test script to reproduce the problem.
- For installation related issue, complete output of `pip install ibm_db` command.

Steps to Reproduce:

First the description of the problem, then the sample script to recreate.

ibm_db.execute: arrays of PYTHON_DECIMAL fail for some valid target column db2-datatypes

CLI allows a PYTHON_DECIMAL value to successfully insert into several different database-column-types, this complies with the "Casting between data types" page of the Db2-documentation. ( https://www.ibm.com/docs/en/db2/11.5?topic=types-casting-between-data )

But python ibm_db limits the target database column datatypes for some unknown reason (it seems unnecessary) and can throw exceptions for inserts that CLI supports, for example CLI0113E. Python ibm_db works fine when binding arrays with the target column datatypes are DECIMAL or DECFLOAT. But fails if the target column data types are (for example) DOUBLE/float, VARCHAR, CHAR etc. Examination of the source code in `ibm_db.c` function `_python_ibm_db_bind_data` shows at least two flaws , one will cause the CLI0113E due to incorrect bracing, another restricts the target-column-datatype and results in `"Binding Error:"` (with no exception detail, because the code has no error handling at this specific location when handling PYTHON_DECIMAL).

1.create a target table in the database named `bind_test_decimal` with the following DDL `create table bind_test_decimal ( col1 float  not null , col2 float  not null )`  .  In this example, we will attempt an array-binding of python_decimal types with ibm_db_execute with the code below.

2. Adjust the code fragment below for your database and credentials and run it. With ibm_db version 3.1.4 you may get `"Binding Error:"`.   If you adjust the database table datatype to VARCHAR(100 ) instead of float , which is a valid target type to cast from decimal, then you may get `CLI0113E`.   If you adjust the column datatype to DECFLOAT or DECIMAL/NUMERIC, the code runs fine.

```
import ibm_db
import sys 
from decimal import *



# Adjust for your environment
schema="..."
tablename="bind_test_decimal"
authid=schema
token="..."
databases_list=["..."]    

# Prereq: before running this script, use the following DDL on your database to match above schema.
# drop table if exists bind_test_decimal 
# create table bind_test_decimal(col1 decimal(29,3) not null, col2 decimal(29,3) not null)


# When the script is successful, 10 rows get inserted into table bind_test_decimal.

getcontext().prec=3

# the insert statement does not change so initialise it outside the loop(s)
insert_sql = f"INSERT INTO {schema}.{tablename} ( col1 , col2 ) VALUES (? ,?)"


# sample data 
col1_array=list(map(Decimal, '1.99 2.99 3.10 4.555 5.009 6.001 7.1 8.81 9.999 10.111'.split()))

col2_array=list(map(Decimal, '0.99 -2.99 -3.40 -4.550 -5.109 -6.002 7.2 -8.82 -9.999 10.10'.split()))

array_length=len(col1_array)


for i in range(1):
    for db in databases_list:
        print(f"Database {db} Iteration {i}\n")
        conn = ibm_db.connect( db, authid, token )
        if not conn:
            print(f"Error failed to connect to database {db} :{ibm_db.conn_errormsg()}\n")
            sys.exit(1)
    
        ibm_db.autocommit(conn, ibm_db.SQL_AUTOCOMMIT_ON)
        spare_hstmt = ibm_db.exec_immediate(conn, f"DELETE FROM {schema}.{tablename}")
        ibm_db.free_result(spare_hstmt)
        spare_hstmt=None
    
        insert_stmt=None
        insert_stmt = ibm_db.prepare(conn, insert_sql, {ibm_db.SQL_ATTR_CURSOR_TYPE: ibm_db.SQL_CURSOR_STATIC})
    
        mybool=False
        mybool = ibm_db.set_option(insert_stmt, {ibm_db.SQL_ATTR_PARAM_BIND_TYPE : ibm_db.SQL_PARAM_BIND_BY_COLUMN}, 0)
        if not mybool:
            print(f"Database {db} Iteration {i} ,Failed to set array columnwise binding option: {ibm_db.stmt_errormsg()}\n")
            sys.exit(1)

        mybool=False
        mybool=ibm_db.set_option(insert_stmt, {ibm_db.SQL_ATTR_PARAMSET_SIZE : array_length}, 0)
        if not mybool:
            print(f"Database {db} Iteration {i} , Failed to set array length: {ibm_db.stmt_errormsg()}\n")
            sys.exit(1)
    
        try:
            mybool=False
            mybool=ibm_db.execute(insert_stmt, ( col1_array, col2_array))
    
        except Exception as e:
            print(f"Database {db} Iteration {i}: \n{ibm_db.stmt_errormsg()}")
            raise e

        ibm_db.free_result(insert_stmt)
        ibm_db.close(conn)
        conn=None


sys.exit(0)

```

**Comment by Earammak** [comment]
@imavo, It will work, when you give input array directly, instead of using map function. 
I tried giving direct input array for float and varchar and there is no issue.
May i know purpose you was tried in that format.

Thanks

**Comment by imavo** [comment]
Yes it works without the map() in the simplest case with a simple string of decimals, but the real 
case had objects that I wanted to transform via map(). Why should'nt it work with map()? It still produces a list of decimals, i think.
